rfc-drafts/must_not_await_lint.md

```diff
@@ -0,0 +1,151 @@
+
+- Any RAII guard might possibly create unintended behavior if held across an await boundary.
+
+This lint will enable the compiler to warn the user that the generated MIR could produce unforeseen side effects. Some examples of this are:
+
+- [`std::sync::MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)
+- [`tracing::span::Entered`](https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html)
+
+This will be a best effort lint to signal the user about unintended side-effects of using certain types across an await boundary.
+
+# Reference-level explanation
+
+The `must_not_await` attribute is used to issue a diagnostic warning when a value is not "used". It can be applied to user-defined composite types (structs, enums and unions), functions and traits.
+
+The `must_not_await` attribute may include a message by using the [`MetaNameValueStr`] syntax such as `#[must_not_await = "example message"]`.  The message will be given alongside the warning.
+
+When used on a user-defined composite type, if the [expression] of an [expression statement] has this type and is used across an await point, then this lint is violated.
+
+
+```rust
+#[must_not_await = "Your error message here"]
+struct MyStruct {}
+
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+When used on a function, if the [expression] of an [expression statement] is a [call expression] to that function, and the expression is held across an await point, this lint is violated.
+
+```rust
+#[must_not_await]
+fn foo() -> i32 { 5i32 }
+
+async fn foo() {
+  let bar = foo();
+  my_async_op.await;
+  println!("{:?}", bar);
+}
+```
+
+When used on a [trait declaration], a [call expression] of an [expression statement] to a function that returns an [impl trait] of that trait and if the value is held across an await point, the lint is violated.
+
+```rust
+trait Trait {
+    #[must_not_await]
+    fn foo(&self) -> i32;
+}
+
+impl Trait for i32 {
+    fn foo(&self) -> i32 { 0i32 }
+}
+
+async fn foo() {
+  let bar = 5i32.foo();
+  my_async_op.await;
+  println!("{:?}", bar);
+}
+```
+
+When used on a function in a trait implementation, the attribute does nothing.
+
+[`MetaNameValueStr`]: https://doc.rust-lang.org/reference/attributes.html#meta-item-attribute-syntax
+[expression]: https://doc.rust-lang.org/reference/expressions.html
+[expression statement]: https://doc.rust-lang.org/reference/statements.html#expression-statements
+[call expression]: https://doc.rust-lang.org/reference/expressions/call-expr.html
+[trait declaration]: https://doc.rust-lang.org/reference/items/traits.html
+[impl trait]: https://doc.rust-lang.org/reference/types/impl-trait.html
+
+### Auto trait vs attribute
```

Comment:
I'm not sure this belongs in the reference section. Is this an open question, or does it belong in the rational section?

Reply:
This answers the question of why we chose to go the attribute route instead of creating an auto-trait. I think this should be in the rationale section.